pkg/solana/txm/pendingtx.go

```diff
@@ -138,28 +196,270 @@ func (c *pendingTxContext) ListAll() []solana.Signature {
-	return c.Remove(sig)
+func (c *pendingTxContext) OnConfirmed(sig solana.Signature) (string, error) {
+	c.lock.RLock()
+	// validate if sig exists
+	id, sigExists := c.sigToID[sig]
+	if !sigExists {
+		c.lock.RUnlock()
+		return id, errors.New("signature does not exist")
+	}
+	// Check if transaction already in confirmed state
+	if tx, exists := c.confirmedTxs[id]; exists && tx.state == Confirmed {
+		c.lock.RUnlock()
+		return id, ErrAlreadyInExpectedState
+	}
+	// Transactions should only move to confirmed from broadcasted/processed
+	if _, exists := c.broadcastedTxs[id]; !exists {
+		c.lock.RUnlock()
+		return id, errors.New("id does not exist in broadcasted map")
+	}
+	c.lock.RUnlock()
+
+	// upgrade to write lock if id exists
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	id, sigExists = c.sigToID[sig]
+	if !sigExists {
+		return id, errors.New("signature does not exist")
+	}
+	if _, exists := c.broadcastedTxs[id]; !exists {
+		return id, errors.New("id does not exist in broadcasted map")
+	}
+	// call cancel func + remove from map to stop the retry/bumping cycle for this transaction
+	if cancel, exists := c.cancelBy[id]; exists {
+		cancel() // cancel context
+		delete(c.cancelBy, id)
+	}
+	tx := c.broadcastedTxs[id]
+	// update tx state to Confirmed
+	tx.state = Confirmed
+	// move tx to confirmed map
+	c.confirmedTxs[id] = tx
+	// remove tx from broadcasted map
+	delete(c.broadcastedTxs, id)
+	return id, nil
 }
 
-func (c *pendingTxContext) OnError(sig solana.Signature, _ int) uuid.UUID {
-	return c.Remove(sig)
+func (c *pendingTxContext) OnFinalized(sig solana.Signature, retentionTimeout time.Duration) (string, error) {
+	// check if already cancelled
+	c.lock.RLock()
+	id, sigExists := c.sigToID[sig]
+	if !sigExists {
+		c.lock.RUnlock()
+		return id, errors.New("signature does not exist")
+	}
+	// Allow transactions to transition from broadcasted, processed, or confirmed state in case there are delays between status checks
+	var broadcastedExists, confirmedExists bool
+	_, broadcastedExists = c.broadcastedTxs[id]
+	_, confirmedExists = c.confirmedTxs[id]
+	if !broadcastedExists && !confirmedExists {
+		c.lock.RUnlock()
+		return id, errors.New("id does not exist in broadcasted or confirmed map")
+	}
+	c.lock.RUnlock()
+
+	// upgrade to write lock if id exists
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if _, exists := c.sigToID[sig]; !exists {
+		return id, errors.New("signature does not exist")
+	}
+	var tx, tempTx pendingTx
+	if tempTx, broadcastedExists = c.broadcastedTxs[id]; broadcastedExists {
+		tx = tempTx
+	}
+	if tempTx, confirmedExists = c.confirmedTxs[id]; confirmedExists {
+		tx = tempTx
+	}
+	if !broadcastedExists && !confirmedExists {
+		return id, errors.New("id does not exist in broadcasted or confirmed map")
+	}
+	// call cancel func + remove from map to stop the retry/bumping cycle for this transaction
+	// cancel is expected to be called and removed when tx is confirmed but checked here too in case state is skipped
```

Comment:
I added this just in case there's some transient issue with RPCs and/or maybe a slow confirmation poll period config. There could be a scenario where we read the state of a tx as `Processed` in one run loop but for one reason or another the next time we check it's already `Finalized`.